src/digests.rs

```diff
@@ -8,34 +9,155 @@ use crate::errors::{DecodeError, DecodeOwnedError};
+///     Foo = 0x01,
+///     Bar = 0x02,
+/// }
+///
+/// impl TryFrom<u64> for MyCodeTable {
+///     type Error = String;
+///
+///     fn try_from(raw: u64) -> Result<Self, Self::Error> {
+///         match raw {
+///             0x01 => Ok(Self::Foo),
+///             0x02 => Ok(Self::Bar),
+///             _ => Err("invalid code".to_string()),
+///         }
+///     }
+/// }
+///
+/// impl From<MyCodeTable> for u64 {
+///     fn from(code: MyCodeTable) -> Self {
+///         code as u64
+///     }
+/// }
+///
+/// #[derive(Clone, Debug)]
+/// struct SameHash;
+/// impl SameHash {
+///     pub const CODE: MyCodeTable = MyCodeTable::Foo;
+///     /// Hash some input and return the sha1 digest.
+///     pub fn digest(_data: &[u8]) -> MultihashGeneric<MyCodeTable> {
+///         let digest = b"alwaysthesame";
+///         wrap(Self::CODE, digest)
+///     }
+/// }
+///
+/// let my_hash = SameHash::digest(b"abc");
+/// assert_eq!(my_hash.digest(), b"alwaysthesame");
+/// ```
+///
+/// This mechanism can also be used if you want to extend the existing code table
+///
+/// # Example
+///
+/// ```
+/// use multihash::Code;
+/// use std::convert::TryFrom;
+///
+/// #[derive(Debug, PartialEq)]
+/// enum ExtendedCode {
+///     Foo,
+///     Bar,
+///     NormalCode(Code),
+/// }
+///
+/// impl TryFrom<u64> for ExtendedCode {
+///     type Error = String;
+///
+///     /// Return the `Code` based on the integer value
+///     fn try_from(raw: u64) -> Result<Self, Self::Error> {
+///         match raw {
+///             0x01 => Ok(Self::Foo),
+///             0x02 => Ok(Self::Bar),
+///             // Fallback to the default values
+///             _ => match Code::try_from(raw) {
+///                 Ok(code) => Ok(Self::NormalCode(code)),
+///                 Err(_) => Err("invalid code".to_string()),
+///             }, //_ => Err("invalid code".to_string()),
+///         }
+///     }
+/// }
+///
+/// impl From<ExtendedCode> for u64 {
+///     fn from(code: ExtendedCode) -> Self {
+///         match code {
+///             ExtendedCode::Foo => 0x01,
+///             ExtendedCode::Bar => 0x02,
+///             ExtendedCode::NormalCode(normal_code) => normal_code.into(),
+///         }
+///     }
+/// }
+///
+/// impl TryFrom<ExtendedCode> for Code {
+///     type Error = String;
+///
+///     fn try_from(extended: ExtendedCode) -> Result<Self, Self::Error> {
+///         match extended {
+///             ExtendedCode::NormalCode(code) => Ok(code),
+///             _ => Err("Not a default code".to_string()),
+///         }
+///     }
+/// }
+///
+/// assert_eq!(ExtendedCode::try_from(0x02).unwrap(), ExtendedCode::Bar);
+/// assert_eq!(
+///     ExtendedCode::try_from(0x12).unwrap(),
+///     ExtendedCode::NormalCode(Code::Sha2_256)
+/// );
+/// assert_eq!(
+///     Code::try_from(ExtendedCode::try_from(0x12).unwrap()).unwrap(),
+///     Code::Sha2_256
+/// );
+/// ```
 #[derive(Clone)]
-pub struct Multihash {
+pub struct MultihashGeneric<T> {
```

Comment:
This should have a trait bound of `TryFrom<u64>` given how much this is being relied upon in the other impls.